README.md
# ZMQ::FFI [![Build Status](https://api.travis-ci.org/zeromq/perlzmq.svg?branch=master)](https://travis-ci.org/zeromq/perlzmq)

## version agnostic Perl bindings for ØMQ using ffi ##

ZMQ::FFI exposes a high level, transparent, OO interface to zeromq independent of the underlying libzmq version.  Where semantics differ, it will dispatch to the appropriate backend for you.  As it uses ffi, there is no dependency on XS or compilation.

ZMQ::FFI is implemented using [FFI::Platypus](https://github.com/plicease/FFI-Platypus).

### EXAMPLES ###

#### send/recv ####
```perl
use 5.012;
use ZMQ::FFI qw(ZMQ_REQ ZMQ_REP);

my $endpoint = "ipc://zmq-ffi-$$";
my $ctx      = ZMQ::FFI->new();

my $s1 = $ctx->socket(ZMQ_REQ);
$s1->connect($endpoint);

my $s2 = $ctx->socket(ZMQ_REP);
$s2->bind($endpoint);

$s1->send('ohhai');

say $s2->recv();
# ohhai
```

#### pub/sub ####
```perl
use 5.012;
use ZMQ::FFI qw(ZMQ_PUB ZMQ_SUB);
use Time::HiRes q(usleep);

my $endpoint = "ipc://zmq-ffi-$$";
my $ctx      = ZMQ::FFI->new();

my $s = $ctx->socket(ZMQ_SUB);
my $p = $ctx->socket(ZMQ_PUB);

$s->connect($endpoint);
$p->bind($endpoint);

# all topics
{
    $s->subscribe('');

    until ($s->has_pollin) {
        # compensate for slow subscriber
        usleep 100_000;
        $p->send('ohhai');
    }

    say $s->recv();
    # ohhai

    $s->unsubscribe('');
}

# specific topics
{
    $s->subscribe('topic1');
    $s->subscribe('topic2');

    until ($s->has_pollin) {
        usleep 100_000;
        $p->send('topic1 ohhai');
        $p->send('topic2 ohhai');
    }

    while ($s->has_pollin) {
        say join ' ', $s->recv();
        # topic1 ohhai
        # topic2 ohhai
    }
}
```

#### multipart ####
```perl
use 5.012;
use ZMQ::FFI qw(ZMQ_DEALER ZMQ_ROUTER);

my $endpoint = "ipc://zmq-ffi-$$";
my $ctx      = ZMQ::FFI->new();

my $d = $ctx->socket(ZMQ_DEALER);
$d->set_identity('dealer');

my $r = $ctx->socket(ZMQ_ROUTER);

$d->connect($endpoint);
$r->bind($endpoint);

$d->send_multipart([qw(ABC DEF GHI)]);

say join ' ', $r->recv_multipart;
# dealer ABC DEF GHI
```

#### nonblocking ####
```perl
use 5.012;
use ZMQ::FFI qw(ZMQ_PUSH ZMQ_PULL);
use AnyEvent;
use EV;

my $endpoint = "ipc://zmq-ffi-$$";
my $ctx      = ZMQ::FFI->new();
my @messages = qw(foo bar baz);


my $pull = $ctx->socket(ZMQ_PULL);
$pull->bind($endpoint);

my $fd = $pull->get_fd();

my $recv = 0;
my $w = AE::io $fd, 0, sub {
    while ( $pull->has_pollin ) {
        say $pull->recv();
        # foo, bar, baz

        $recv++;
        if ($recv == 3) {
            EV::break();
        }
    }
};


my $push = $ctx->socket(ZMQ_PUSH);
$push->connect($endpoint);

my $sent = 0;
my $t;
$t = AE::timer 0, .1, sub {
    $push->send($messages[$sent]);

    $sent++;
    if ($sent == 3) {
        undef $t;
    }
};

EV::run();
```

#### specifying versions ####
```perl
use ZMQ::FFI;

# 2.x context
my $ctx = ZMQ::FFI->new( soname => 'libzmq.so.1' );
my ($major, $minor, $patch) = $ctx->version;

# 3.x context
my $ctx = ZMQ::FFI->new( soname => 'libzmq.so.3' );
my ($major, $minor, $patch) = $ctx->version;
```

### INSTALL ###

    cpanm -v ZMQ::FFI

### BUILD ###

A docker image is provided with a pre-configured testing environment. To test the module:

    ./docker-run dzil test

To build a dist tarball:

    ./docker-run dzil build

To clean build artifacts:

    ./docker-run dzil clean

Tests will run against every stable version of zeromq as well as master.  If you would like an interactive shell inside the container run `./docker-shell`

If you would prefer a native local setup refer to the Dockerfile and translate the setup steps accordingly for your distribution/platform (I personally use the docker container, and this is also how tests run under Travis).

### DOCUMENTATION ###

https://metacpan.org/module/ZMQ::FFI



Changes

{{$NEXT}}

1.19      2023-08-04 11:16:25+00:00 UTC

1.18      2022-03-21 10:48:44+00:00 UTC

    - New maintainer @ghenry - welcome and thank you!
    - Add MSWin32 support (GH#47 - thanks @zmughal!)
    - Add zmq_socket_monitor support (GH#46 - thanks @potatogim!)
    - Add z85_encoding support (GH#43 - thanks @juddtaylor!)
    - Document contributors
    - Update repository info. Now officially part of the zeromq project at
      https://github.com/zeromq/perlzmq
    - Have examples and scripts use 5.12 as minimum version, which enables
      strict by default

1.17      2019-03-14 04:05:15+00:00 UTC

1.16_01   2019-03-13 03:40:02+00:00 UTC (TRIAL RELEASE)

    - Need FFI::Platypus specified as a configure dependency
      (GH#42 - thanks @eserte)

1.15      2019-03-03 10:14:00+00:00 UTC

    - Bump FFI::Platypus dependency version to 0.86
      https://github.com/Perl5-FFI/FFI-Platypus/issues/117
    - Fix t/device.t hang, seems to be same cause/fix as GH#37
    - Update ZMQ::FFI::Constants pod to also list libzmq as one of the repos
      used
    - Don't prune constants generation script from release tarball

1.14      2019-02-28 18:49:49+00:00 UTC

    - The ZMQ::FFI::Constants generation script wasn't using stable releases
      tagged in the libzmq main repo, this is fixed
    - FFI::TinyCC is no longer an author dep and has been replaced by a docker
      based solution for generating ZMQ::FFI::Constants, including calculating
      zmq_msg_t sizes
    - Fix potential memleak around socket cleanup (GH#33 - thanks @rhrhunter!)
    - Use string literals for matching option types (GH#30 - thanks @bbkr)

1.12      2019-02-26 05:16:15-05:00 America/New_York

    - Fix t/proxy.t hang (GH#37)

1.11      2016-01-10 19:27:05-05:00 America/New_York

    - Fix bareword error in closed_socket.t (GH#27 - thanks @rhrhunter)

1.10      2016-01-09 15:50:59-05:00 America/New_York

    - support importing zmq constants directly through ZMQ::FFI (GH#24)
    - don't execute operations on closed sockets and emit appropriate warnings
      (GH#23)

1.09      2015-12-09 04:55:20-06:00 America/Chicago

    - add more robust cleanup handling for ooo global destruction scenarios.

1.08      2015-11-07 10:47:10-06:00 America/Chicago

    - Fix cleanup handling when using Perl threads

1.07      2015-09-19 12:38:32-05:00 America/Chicago

    - Add 4.1 to xt/test_versions.sh
    - Add missing 4.1 stable constants. Update constants generation script
      to handle hex constant values.
    - Add workaround for hang in device.t and proxy.t

1.06      2015-03-28 11:30:15-05:00 America/Chicago

    - Update FreeBSD check

1.05_02   2015-03-28 03:31:38-05:00 America/Chicago

1.05_01   2015-03-28 02:52:01-05:00 America/Chicago

1.04      2015-03-24 03:02:20-05:00 America/Chicago

    - require FFI::Platypus 0.33 which fixes potential segfaults during global
      destruction (GH#19)
    - don't install on unthreaded BSD perls (GH#13)
    - more performance improvements, incl. 30% faster recv rate in some tests
      see 83f0013, f834fe9, and 394e164
    - use a default linger of 0 (default used by czmq/jzmq)
    - don't clobber user linger value (GH#18)
    - properly localize redefines in tests
    - fix cleanup of ctx/sockets created in forked children

1.03      2015-03-16 04:47:24-05:00 America/Chicago

    - add alternative (non-exceptional) error handling semantics
    - fix socket files being left around after device.t and proxy.t
    - misc doc reformatting and rewording
    - don't use EV in fd.t

1.02      2015-03-12 04:51:14-05:00 America/Chicago

    - fix cleanup handling of inherited contexts/sockets in forked childen
    - don't include datetime in generated Constants module
    - fix pointer pack warning in ZMQ2 backend

1.01      2015-03-11 00:50:39-05:00 America/Chicago

    - Require Math::BigInt 1.997 or newer to resolve integer overflow bug
      (GH#14 - thanks @plicease!)
    - Perl v5.10 is now the official minimum version supported

1.00      2015-03-09 00:54:41-05:00 America/Chicago

    - switch to FFI::Platypus on the backend. FFI::Platypus provides the
      ability to bind ffi functions as first class Perl xsubs, resulting in
      dramatic performance gains.
    - optimize the Perl code in general, especially the send/recv hot spots
    - require Class::XSAccessor, which substantially improves Moo accessor
      performance
    - don't test against dev Perl on travis for now, seems to be busted

0.19      2015-03-04 01:42:16-06:00 America/Chicago

    - use dzil FFI::CheckLib plugin to properly report NA on CPAN tester boxes
      without libzmq
    - test against dev perl in addition to stable on travis

0.18      2015-02-25 07:30:20-06:00 America/Chicago

    - linger default has changed back to -1 for libzmq all versions
    - add travis tests for Perl 5.20
    - $! will not be localized by default when stringified in Perl 5.22 and
      beyond, and needs to be explicitly localized in the tests (GH#12)

0.17      2014-11-08 22:31:25-06:00 America/Chicago

    - add zmq_device and zmq_proxy functionality
      (GH#10, GH#11 - thanks @semifor)
    - add libzmq.so.4 to list of sonames checked
    - linger default is 2000 circa libzmq 4.2

0.16      2014-09-13 17:20:05-05:00 America/Chicago

    - generate zmq_msg_t size, don't hardcode it
      (GH#9 - thanks @parth-j-gandhi!)
    - test against libzmq dev repo in addition to stable

0.15      2014-08-15 20:39:39 America/Chicago

    - Apply flags correctly in multipart sends (GH#8 - thanks @shripchenko)

0.14      2014-07-06 00:39:20 America/Chicago

    - add disconnect POD

0.13      2014-07-05 17:03:08 America/Chicago

    - add zmq_disconnect and zmq_unbind bindings (GH#7)

0.12      2014-03-29 17:48:45 America/Chicago

    - fix binary/string option handling (e.g. ZMQ_LAST_ENDPOINT)

0.11      2014-02-17 19:50:14 America/Chicago

    - works on OS X now (GH#6 - thanks @wjackson!)
    - mucho refactor and code reorg


lib/ZMQ/FFI.pm
package ZMQ::FFI;

# ABSTRACT: version agnostic Perl bindings for zeromq using ffi

use strict;
use warnings;

use ZMQ::FFI::Util qw(zmq_soname zmq_version valid_soname);
use Carp;

use Import::Into;

sub import {
    my ($pkg, @import_args) = @_;

    my $target = caller;
    ZMQ::FFI::Constants->import::into($target, @import_args);
}

sub new {
    my ($self, %args) = @_;

    if ($args{soname}) {
        unless ( valid_soname($args{soname}) ) {
            die "Failed to load '$args{soname}', is it on your loader path?";
        }
    }
    else {
        $args{soname} = zmq_soname( die => 1 );
    }

    my ($major, $minor) = zmq_version($args{soname});

    if ($major == 2) {
        require ZMQ::FFI::ZMQ2::Context;
        return ZMQ::FFI::ZMQ2::Context->new(%args);
    }
    elsif ($major == 3) {
        require ZMQ::FFI::ZMQ3::Context;
        return ZMQ::FFI::ZMQ3::Context->new(%args);
    } 
    else {
	if ($major == 4 and $minor == 0) {
            require ZMQ::FFI::ZMQ4::Context;
            return ZMQ::FFI::ZMQ4::Context->new(%args);
        }
        else {
            require ZMQ::FFI::ZMQ4_1::Context;
            return ZMQ::FFI::ZMQ4_1::Context->new(%args);
        }
    }
}

1;

__END__

=head1 SYNOPSIS

    #### send/recv ####

    use 5.012;
    use ZMQ::FFI qw(ZMQ_REQ ZMQ_REP);

    my $endpoint = "ipc://zmq-ffi-$$";
    my $ctx      = ZMQ::FFI->new();

    my $s1 = $ctx->socket(ZMQ_REQ);
    $s1->connect($endpoint);

    my $s2 = $ctx->socket(ZMQ_REP);
    $s2->bind($endpoint);

    $s1->send('ohhai');

    say $s2->recv();
    # ohhai


    #### pub/sub ####

    use 5.012;
    use ZMQ::FFI qw(ZMQ_PUB ZMQ_SUB);
    use Time::HiRes q(usleep);

    my $endpoint = "ipc://zmq-ffi-$$";
    my $ctx      = ZMQ::FFI->new();

    my $s = $ctx->socket(ZMQ_SUB);
    my $p = $ctx->socket(ZMQ_PUB);

    $s->connect($endpoint);
    $p->bind($endpoint);

    # all topics
    {
        $s->subscribe('');

        until ($s->has_pollin) {
            # compensate for slow subscriber
            usleep 100_000;
            $p->send('ohhai');
        }

        say $s->recv();
        # ohhai

        $s->unsubscribe('');
    }

    # specific topics
    {
        $s->subscribe('topic1');
        $s->subscribe('topic2');

        until ($s->has_pollin) {
            usleep 100_000;
            $p->send('topic1 ohhai');
            $p->send('topic2 ohhai');
        }

        while ($s->has_pollin) {
            say join ' ', $s->recv();
            # topic1 ohhai
            # topic2 ohhai
        }
    }


    #### multipart ####

    use 5.012;
    use ZMQ::FFI qw(ZMQ_DEALER ZMQ_ROUTER);

    my $endpoint = "ipc://zmq-ffi-$$";
    my $ctx      = ZMQ::FFI->new();

    my $d = $ctx->socket(ZMQ_DEALER);
    $d->set_identity('dealer');

    my $r = $ctx->socket(ZMQ_ROUTER);

    $d->connect($endpoint);
    $r->bind($endpoint);

    $d->send_multipart([qw(ABC DEF GHI)]);

    say join ' ', $r->recv_multipart;
    # dealer ABC DEF GHI


    #### nonblocking ####

    use 5.012;
    use ZMQ::FFI qw(ZMQ_PUSH ZMQ_PULL);
    use AnyEvent;
    use EV;

    my $endpoint = "ipc://zmq-ffi-$$";
    my $ctx      = ZMQ::FFI->new();
    my @messages = qw(foo bar baz);


    my $pull = $ctx->socket(ZMQ_PULL);
    $pull->bind($endpoint);

    my $fd = $pull->get_fd();

    my $recv = 0;
    my $w = AE::io $fd, 0, sub {
        while ( $pull->has_pollin ) {
            say $pull->recv();
            # foo, bar, baz

            $recv++;
            if ($recv == 3) {
                EV::break();
            }
        }
    };


    my $push = $ctx->socket(ZMQ_PUSH);
    $push->connect($endpoint);

    my $sent = 0;
    my $t;
    $t = AE::timer 0, .1, sub {
        $push->send($messages[$sent]);

        $sent++;
        if ($sent == 3) {
            undef $t;
        }
    };

    EV::run();


    #### specifying versions ####

    use ZMQ::FFI;

    # 2.x context
    my $ctx = ZMQ::FFI->new( soname => 'libzmq.so.1' );
    my ($major, $minor, $patch) = $ctx->version;

    # 3.x context
    my $ctx = ZMQ::FFI->new( soname => 'libzmq.so.3' );
    my ($major, $minor, $patch) = $ctx->version;


=head1 DESCRIPTION

ZMQ::FFI exposes a high level, transparent, OO interface to zeromq independent
of the underlying libzmq version.  Where semantics differ, it will dispatch to
the appropriate backend for you.  As it uses ffi, there is no dependency on XS
or compilation.

As of 1.00 ZMQ::FFI is implemented using L<FFI::Platypus>. This version has
substantial performance improvements and you are encouraged to use 1.00 or
newer.

=head1 CONTEXT API

=head2 new

    my $ctx = ZMQ::FFI->new(%options);

returns a new context object, appropriate for the version of
libzmq found on your system. It accepts the following optional attributes:

=head3 options

=over 4

=item threads

zeromq thread pool size. Default: 1

=item max_sockets

I<requires zmq E<gt>= 3.x>

max number of sockets allowed for context. Default: 1024

=item soname

    ZMQ::FFI->new( soname => '/path/to/libzmq.so' );
    ZMQ::FFI->new( soname => 'libzmq.so.3' );

specify the libzmq library name to load.  By default ZMQ::FFI will first try
the generic soname for the system, then the soname for each version of zeromq
(e.g. libzmq.so.3). C<soname> can also be the path to a particular libzmq so
file

It is technically possible to have multiple contexts of different versions in
the same process, though the utility of doing such a thing is dubious

=back

=head2 version

    my ($major, $minor, $patch) = $ctx->version();

return the libzmq version as the list C<($major, $minor, $patch)>

=head2 get

I<requires zmq E<gt>= 3.x>

    my $threads = $ctx->get(ZMQ_IO_THREADS)

get a context option value

=head2 set

I<requires zmq E<gt>= 3.x>

    $ctx->set(ZMQ_MAX_SOCKETS, 42)

set a context option value

=head2 socket

    my $socket = $ctx->socket(ZMQ_REQ)

returns a socket of the specified type. See L</SOCKET API> below

=head2 proxy

    $ctx->proxy($frontend, $backend);

    $ctx->proxy($frontend, $backend, $capture);

sets up and runs a C<zmq_proxy>. For zmq 2.x this will use a C<ZMQ_STREAMER>
device to simulate the proxy. The optional C<$capture> is only supported for
zmq E<gt>= 3.x however

=head2 device

I<zmq 2.x only>

    $ctx->device($type, $frontend, $backend);

sets up and runs a C<zmq_device> with specified frontend and backend sockets

=head2 destroy

destroy the underlying zmq context. In general you shouldn't have to call this
directly as it is called automatically for you when the object gets reaped

See L</CLEANUP> below

=head1 SOCKET API

The following API is available on socket objects created by C<$ctx-E<gt>socket>.

For core attributes and functions, common across all versions of zeromq,
convenience methods are provided. Otherwise, generic get/set methods are
provided that will work independent of version.

As attributes are constantly being added/removed from zeromq, it is unlikely
the 'static' accessors will grow much beyond the current set.

=head2 version

    my ($major, $minor, $patch) = $socket->version();

same as Context C<version> above

=head2 connect

    $socket->connect($endpoint);

does socket connect on the specified endpoint

=head2 disconnect

I<requires zmq E<gt>= 3.x>

    $socket->disconnect($endpoint);

does socket disconnect on the specified endpoint

=head2 bind

    $socket->bind($endpoint);

does socket bind on the specified endpoint

=head2 unbind

I<requires zmq E<gt>= 3.x>

    $socket->unbind($endpoint);

does socket unbind on the specified endpoint

=head2 get_linger, set_linger

    my $linger = $socket->get_linger();

    $socket->set_linger($millis);

get or set the socket linger period. Default: 0 (no linger)

See L</CLEANUP> below

=head2 get_identity, set_identity

    my $ident = $socket->get_identity();

    $socket->set_identity($ident);

get or set the socket identity for request/reply patterns

=head2 get_fd

    my $fd = $socket->get_fd();

get the file descriptor associated with the socket

=head2 get

    my $option_value = $socket->get($option_name, $option_type);

    my $linger = $socket->get(ZMQ_LINGER, 'int');

generic method to get the value for any socket option. C<$option_type> is the
type associated with C<$option_value> in the zeromq API (C<zmq_getsockopt> man
page)

=head2 set

    $socket->set($option_name, $option_type, $option_value);

    $socket->set(ZMQ_IDENTITY, 'binary', 'foo');

generic method to set the value for any socket option.  C<$option_type> is the
type associated with C<$option_value> in the zeromq API (C<zmq_setsockopt> man
page)

=head2 subscribe

    $socket->subscribe($topic);

add C<$topic> to the subscription list

=head2 unsubscribe

    $socket->unsubscribe($topic);

remove C<$topic> from the subscription list

=head2 send

    $socket->send($msg);

    $socket->send($msg, $flags);

sends a message using the optional flags

=head2 send_multipart

    $socket->send($parts_aref);

    $socket->send($parts_aref, $flags);

given an array ref of message parts, sends the multipart message using the
optional flags. ZMQ_SNDMORE semantics are handled for you

=head2 recv

    my $msg = $socket->recv();

    my $msg = $socket->recv($flags);

receives a message using the optional flags

=head2 recv_multipart

    my @parts = $socket->recv_multipart();

    my @parts = $socket->recv_multipart($flags);

receives a multipart message, returning an array of parts. ZMQ_RCVMORE
semantics are handled for you

=head2 has_pollin, has_pollout

    while ( $socket->has_pollin ) { ... }

checks ZMQ_EVENTS for ZMQ_POLLIN and ZMQ_POLLOUT respectively, and returns
true/false depending on the state

=head2 close

close the underlying zmq socket. In general you shouldn't have to call this
directly as it is called automatically for you when the object gets reaped

See L</CLEANUP> below

=head2 die_on_error

    $socket->die_on_error(0);

    $socket->die_on_error(1);

controls whether error handling should be exceptional or not. This is set to
true by default. See L</ERROR HANDLING> below

=head2 has_error

returns true or false depending on whether the last socket operation had an
error. This is really just an alias for C<last_errno>

=head2 last_errno

returns the system C<errno> set by the last socket operation, or 0 if there
was no error

=head2 last_strerror

returns the human readable system error message associated with the socket
C<last_errno>

=head1 CLEANUP

With respect to cleanup C<ZMQ::FFI> follows either the L<zeromq guide|http://zguide.zeromq.org/page:all#Making-a-Clean-Exit>
recommendations or the behavior of other zmq bindings.
That is:

=over 4

=item * it uses 0 linger by default (this is the default used by L<czmq|https://github.com/zeromq/czmq> and L<jzmq|https://github.com/zeromq/jzmq>)

=item * during object destruction it will call close/destroy for you

=item * it arranges the reference hierarchy such that sockets will be properly
      cleaned up before their associated contexts

=item * it detects fork/thread situations and ensures sockets/contexts are only
      cleaned up in their originating process/thread

=item * it guards against double closes/destroys

=back

Given the above you're probably better off letting C<ZMQ::FFI> handle cleanup
for you. But if for some reason you want to do explicit cleanup yourself you
can. All the below will accomplish the same thing:

    # implicit cleanup
    {
        my $context = ZMQ::FFI->new();
        my $socket  = $ctx->socket($type);
        ...
        # close/destroy called in destructors at end of scope
    }

    # explicit cleanup
    $socket->close();
    $context->destroy();

    # ditto
    undef $socket;
    undef $context;

Regarding C<linger>, you can always set this to a value you prefer if
you don't like the default. Once set the new value will be used when the
socket is subsequently closed (either implicitly or explicitly):

    $socket->set_linger(-1); # infinite linger
                             # $context->destroy will block forever
                             # (or until all pending messages have been sent)

=head1 ERROR HANDLING

By default, ZMQ::FFI checks the return codes of underlying zmq functions for
you, and in the case of an error it will die with the human readable system
error message.

    $ctx->socket(-1);
    # dies with 'zmq_socket: Invalid argument'

Usually this is what you want, but not always. Some zmq operations can return
errors that are not fatal and should be handled. For example using
C<ZMQ_DONTWAIT> with send/recv can return C<EAGAIN> and simply means try
again, not die.

For situations such as this you can turn off exceptional error handling by
setting C<die_on_error> to 0. It is then for you to check and manage any zmq
errors by checking C<last_errno>:

    use Errno qw(EAGAIN);

    my $ctx = ZMQ::FFI->new();
    my $s   = $ctx->socket(ZMQ_DEALER);
    $s->bind('tcp://*:7200');

    $s->die_on_error(0); # turn off exceptional error handling

    while (1) {
        my $msg = $s->recv(ZMQ_DONTWAIT);

        if ($s->last_errno == EAGAIN) {
            sleep 1;
        }
        elsif ($s->last_errno) {
            die $s->last_strerror;
        }
        else {
            warn "recvd: $msg";
            last;
        }
    }

    $s->die_on_error(1); # turn back on exceptional error handling

=head1 FFI VS XS PERFORMANCE

ZMQ::FFI uses L<FFI::Platypus> on the backend. In addition to a friendly,
usable interface, FFI::Platypus's killer feature is C<attach>. C<attach> makes
it possible to bind ffi functions in memory as first class Perl xsubs. This
results in dramatic performance gains and gives you the flexibility of ffi
with performance approaching that of XS.

Testing indicates FFI::Platypus xsubs are around 30% slower than "real" XS
xsubs. That may sound like a lot, but to put it in perspective that means, for
zeromq, the XS bindings can send 10 million messages 1-2 seconds faster than
the ffi ones.

If you really care about 1-2 seconds over 10 million messages you should be
writing your solution in C anyways. An equivalent C implementation will be
several I<hundred> percent faster or more.

Keep in mind also that the small speed bump you get using XS can easily be
wiped out by crappy and poorly optimized Perl code.

Now that Perl finally has a great ffi interface, it is hard to make the case
to continue using XS. The slight speed bump just isn't worth giving up the
convenience, flexibility, and portability of ffi.

You can find the detailed performance results that informed this section at:
L<https://gist.github.com/calid/17df5bcfb81c83786d6f>

=head1 BUGS

C<ZMQ::FFI> is free as in beer in addition to being free as in speech. While
I've done my best to ensure it's tasty, high quality beer, it probably isn't perfect.
If you encounter problems, or otherwise see room for improvement, please open
an issue (or even better a pull request!) on L<github|https://github.com/zeromq/perlzmq>

=head1 SEE ALSO

=for :list
* L<ZMQ::FFI::Constants>
* L<ZMQ::FFI::Util>
* L<FFI::Platypus>
* L<FFI::Raw>
* L<ZMQ::LibZMQ3>

=head1 CREDITS

Thank you to the following for patches, bug reports, feedback, or suggestions:

Dave Lambley, Graham Ollis, Klaus Ita, Marc Mims, Parth Gandhi, Pawel Pabian,
Robert Hunter, Sergey KHripchenko, Slaven Rezic, Whitney Jackson, pipcet
